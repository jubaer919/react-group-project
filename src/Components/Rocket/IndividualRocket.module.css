.container {
  padding: 1rem 5vw;
  display: flex;
  align-items: flex-start;
  gap: 1rem;
}

.img {
  width: 300px;
}

.info-container {
  text-align: left;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

.info-container h1 {
  padding: 0;
  margin: 0 0 10px 0;
}

.reserve-btn {
  display: inline-block;
  border: none;
  outline: none;
  padding: 8px 16px;
  background-color: blue;
  color: white;
  cursor: pointer;
  margin-top: 10px;
  border-radius: 5px;
}

.cancel-btn {
  display: inline-block;
  border: 1px solid black;
  border-radius: 5px;
  background-color: white;
  color: black;
  cursor: pointer;
  margin-top: 10px;
  padding: 8px 1rem;
  outline: none;
}

.reserved {
  margin-right: 1rem;
  padding: 2px 12px;
  background-color: palegreen;
  color: white;
}
