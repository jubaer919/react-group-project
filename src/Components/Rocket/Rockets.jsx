// import { useSelector } from 'react-redux';
// import { allRocketsInfo } from '../../Redux/Rocket/rocketSlice';
// import IndividualRocket from './IndividualRocket';

// export default function Rockets() {
//   const Rockets = useSelector(allRocketsInfo);
//   return (
//     <div>
//       {Rockets.map((rocket) => <IndividualRocket key={rocket.id} rocket={rocket} />)}
//     </div>
//   );
// }
