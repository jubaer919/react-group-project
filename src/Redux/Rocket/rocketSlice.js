import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  rocket: [],
  status: 'good',
  error: null,
};

const baseURL = 'https://api.spacexdata.com/v4/rockets';

export const fetchRocketInfo = createAsyncThunk('rockets/fetchrocketInfo', async () => {
  const response = await axios.get(baseURL);
  return response.data;
});

const rocketSlice = createSlice({
  name: 'rocket',
  initialState,
  reducers: {
    rocketReserve: (state, action) => {
      const id = action.payload;
      const newState = state.rocket.map((rocket) => {
        if (rocket.id !== id) {
          return rocket;
        }
        return { ...rocket, reserved: true };
      });

      return { ...state, rocket: newState };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRocketInfo.pending, (state) => {
        state.status = 'Rocket info is showing up';
      })
      .addCase(fetchRocketInfo.fulfilled, (state, action) => {
        state.status = 'Rocket info Successfull';
        state.rocket = action.payload.map((rocket) => (
          {
            id: rocket.id,
            name: rocket.name,
            des: rocket.description,
            img: rocket.flickr_images[0],
            reserved: false,
          }
        ));
      })
      .addCase(fetchRocketInfo.rejected, (state, action) => {
        state.status = 'Rocket info rejected';
        state.error = action.error.message;
      });
  },
});

export const { rocketReserve } = rocketSlice.actions;
export default rocketSlice.reducer;
