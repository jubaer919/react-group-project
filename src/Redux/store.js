import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './Rocket/rocketSlice';
import missionReducer from './missions/missions';

const store = configureStore({
  reducer: {
    rocket: rocketReducer,
    missions: missionReducer,
  },
});

export default store;
