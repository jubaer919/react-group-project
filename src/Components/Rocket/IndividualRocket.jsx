import PropType from 'prop-types';
import { useDispatch } from 'react-redux';
import { rocketReserve, cancleReservation } from '../../Redux/Rocket/rocketSlice';
import classes from './IndividualRocket.module.css';

export default function IndividualRocket({
  name, img, des, id, re,
}) {
  const dispatch = useDispatch();
  const reservedHandeler = () => {
    dispatch(rocketReserve(id));
  };

  const cancleHandeler = () => {
    dispatch(cancleReservation(id));
  };

  return (
    <div className={classes.container}>
      <div className={classes['img-container']}>
        <img src={img} alt="rocket" className={classes.img} />
      </div>
      <div className={classes['info-container']}>
        <h1>{ name }</h1>
        <div>
          {re
            && <span className={classes.reserved}>Reserved</span>}
          <span>{des}</span>
        </div>
        <div>
          {!re && <button type="button" className={classes['reserve-btn']} onClick={reservedHandeler}>Reserve Rocket</button>}
          {re && <button type="button" className={classes['cancel-btn']} onClick={cancleHandeler}>Cancell Reservation</button>}
        </div>
      </div>
    </div>
  );
}

IndividualRocket.propTypes = {
  name: PropType.string.isRequired,
  img: PropType.string.isRequired,
  des: PropType.string.isRequired,
  id: PropType.string.isRequired,
  re: PropType.bool.isRequired,
};
