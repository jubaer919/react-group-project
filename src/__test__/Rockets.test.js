import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Rockets from '../Components/Rocket/Rockets';
import store from '../Redux/store';
import { rocketReserve, cancleReservation } from '../Redux/Rocket/rocketSlice';

describe('Missions tests', () => {
  it('render test', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should return an action with type rocket/cancleReservation', () => {
    const data = [
      {
        id: 1,
        rocket_id: 'rock52',
        rocket_name: 'rock 52',
        rocket_type: 'rocket',
        flickr_images: ['image1.png', 'imag2.jpg'],
        description: 'A test rocket',
      },
    ];
    const action = cancleReservation(data);
    expect(action.type).toBe('rocket/cancleReservation');
  });

  it('should return an action with type rocket/rocketReserve', () => {
    const id = 1;
    const action = rocketReserve(id);
    expect(action.type).toBe('rocket/rocketReserve');
  });
});
