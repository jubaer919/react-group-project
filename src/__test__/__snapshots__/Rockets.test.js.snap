// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions tests render test 1`] = `<div />`;
